File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/utilities/IconProvider.kt
package com.eopeter.fluttermapboxnavigation.utilities

import com.eopeter.fluttermapboxnavigation.R
import com.eopeter.fluttermapboxnavigation.models.TripProgressTheme

/**
 * Interface for providing icons and colors based on category/iconId.
 *
 * Implement this interface to customize the icons and colors used in the
 * trip progress panel. This allows for full customization without modifying
 * the core component.
 *
 * Example usage:
 * ```kotlin
 * class MyIconProvider : IconProvider {
 *     override fun getIconResource(iconId: String?, category: String): Int {
 *         return when (iconId?.lowercase() ?: category.lowercase()) {
 *             "custom_icon" -> R.drawable.my_custom_icon
 *             else -> DefaultIconProvider.getIconResource(iconId, category)
 *         }
 *     }
 * }
 * ```
 */
interface IconProvider {
    /**
     * Get the drawable resource ID for an icon.
     *
     * @param iconId Optional specific icon ID (takes precedence over category)
     * @param category The waypoint category (e.g., "checkpoint", "waypoint", "poi")
     * @return Drawable resource ID
     */
    fun getIconResource(iconId: String?, category: String): Int

    /**
     * Get the color for a category.
     *
     * @param category The waypoint category
     * @param theme The current theme configuration
     * @return Color int value
     */
    fun getCategoryColor(category: String, theme: TripProgressTheme): Int
}

/**
 * Default implementation of IconProvider.
 *
 * Provides a standard set of icons for common waypoint types including:
 * - Checkpoints (flag)
 * - Waypoints (pin)
 * - POIs (various types like restaurants, hotels, fuel stations, etc.)
 *
 * Uses Material Design-inspired icons and colors.
 */
object DefaultIconProvider : IconProvider {

    /**
     * Get the drawable resource ID for an icon.
     *
     * Falls back to a pin icon if the iconId/category is not recognized.
     */
    override fun getIconResource(iconId: String?, category: String): Int {
        val id = (iconId ?: category).lowercase()

        return when (id) {
            "flag", "checkpoint" -> R.drawable.ic_flag
            "pin", "waypoint" -> R.drawable.ic_pin
            "scenic", "viewpoint", "photo" -> R.drawable.ic_scenic
            "petrol_station", "petrol", "gas", "fuel" -> R.drawable.ic_petrol_station
            "restaurant", "food", "dining" -> R.drawable.ic_restaurant
            "hotel", "accommodation", "lodging" -> R.drawable.ic_hotel
            "parking", "car_park" -> R.drawable.ic_parking
            "hospital", "medical", "clinic" -> R.drawable.ic_hospital
            "police", "emergency" -> R.drawable.ic_police
            "charging_station", "charging", "ev" -> R.drawable.ic_charging_station
            "attraction", "landmark" -> R.drawable.ic_flag
            "rest_area", "rest_stop" -> R.drawable.ic_parking
            "shop", "shopping" -> R.drawable.ic_pin
            else -> R.drawable.ic_pin
        }
    }

    /**
     * Get the color for a category.
     *
     * Uses the theme's category colors if defined, otherwise falls back to
     * the theme's primary color.
     */
    override fun getCategoryColor(category: String, theme: TripProgressTheme): Int {
        return theme.getCategoryColor(category)
    }
}

File: ios/Classes/IconProvider.swift
import UIKit

/// Protocol for providing icons and colors based on category/iconId.
///
/// Implement this protocol to customize the icons and colors used in the
/// trip progress panel. This allows for full customization without modifying
/// the core component.
///
/// Example usage:
/// ```swift
/// class MyIconProvider: IconProvider {
///     func getIcon(iconId: String?, category: String) -> UIImage? {
///         switch (iconId ?? category).lowercased() {
///         case "custom_icon":
///             return UIImage(named: "my_custom_icon")
///         default:
///             return DefaultIconProvider.shared.getIcon(iconId: iconId, category: category)
///         }
///     }
/// }
/// ```
public protocol IconProvider {
    /// Get the icon for a waypoint.
    ///
    /// - Parameters:
    ///   - iconId: Optional specific icon ID (takes precedence over category)
    ///   - category: The waypoint category (e.g., "checkpoint", "waypoint", "poi")
    /// - Returns: UIImage for the icon, or nil if not found
    func getIcon(iconId: String?, category: String) -> UIImage?

    /// Get the color for a category.
    ///
    /// - Parameters:
    ///   - category: The waypoint category
    ///   - theme: The current theme configuration
    /// - Returns: UIColor for the category
    func getCategoryColor(_ category: String, theme: TripProgressTheme) -> UIColor
}

/// Default implementation of IconProvider.
///
/// Provides a standard set of SF Symbols icons for common waypoint types including:
/// - Checkpoints (flag)
/// - Waypoints (pin)
/// - POIs (various types like restaurants, hotels, fuel stations, etc.)
public class DefaultIconProvider: IconProvider {
    /// Shared instance for convenience.
    public static let shared = DefaultIconProvider()

    private init() {}

    /// Get the icon for a waypoint using SF Symbols.
    ///
    /// Falls back to a pin icon if the iconId/category is not recognized.
    public func getIcon(iconId: String?, category: String) -> UIImage? {
        let id = (iconId ?? category).lowercased()

        let symbolName: String
        switch id {
        case "flag", "checkpoint":
            symbolName = "flag.fill"
        case "pin", "waypoint":
            symbolName = "mappin"
        case "scenic", "viewpoint", "photo":
            symbolName = "camera.fill"
        case "petrol_station", "petrol", "gas", "fuel":
            symbolName = "fuelpump.fill"
        case "restaurant", "food", "dining":
            symbolName = "fork.knife"
        case "hotel", "accommodation", "lodging":
            symbolName = "bed.double.fill"
        case "parking", "car_park":
            symbolName = "car.fill"
        case "hospital", "medical", "clinic":
            symbolName = "cross.fill"
        case "police", "emergency":
            symbolName = "shield.fill"
        case "charging_station", "charging", "ev":
            symbolName = "bolt.fill"
        case "attraction", "landmark":
            symbolName = "flag.fill"
        case "rest_area", "rest_stop":
            symbolName = "car.fill"
        case "shop", "shopping":
            symbolName = "bag.fill"
        default:
            symbolName = "mappin"
        }

        return UIImage(systemName: symbolName)
    }

    /// Get the color for a category.
    ///
    /// Uses the theme's category colors if defined, otherwise falls back to
    /// the theme's primary color.
    public func getCategoryColor(_ category: String, theme: TripProgressTheme) -> UIColor {
        return theme.getCategoryColor(category)
    }
}
